man/man1/secstored.1
.TH SECSTORED 1
.SH NAME
secstored, secuser \- secstore commands
.SH SYNOPSIS
.br
.B secstored
[-R]
[-S servername]
[-s tcp!*!5356]
[-x mountpoint]
.br
.B secuser
[-v]
username
.br
.PP
.SH DESCRIPTION
.PP
.I Secstored
serves requests from
.MR secstore 1 .
The
.B -R
option supplements the password check with a
call to a RADIUS server, for checking hardware
tokens or other validation.
The
.BR -x mountpoint
option specifies an alternative to the default network
.BR /net .
.PP
.I Secuser
is an administrative command that runs on the
secstore machine, normally the authserver,
to create new accounts and
to change status on existing accounts.
It prompts for account information such as
password and expiration date, writing to
.BR \*9/secstore/who/$uid .
The
.B \*9/secstore
directory should be created mode 770 for the userid
or groupid of the secstored process.
.PP
By default,
.I secstored
warns the client if no account exists.
If you prefer to obscure this information, use
.I secuser
to create an account
.BR FICTITIOUS .
.SH FILES
.B \*9/secstore/who/$uid
secstore account name, expiration date, verifier
.br
.B \*9/secstore/store/$uid/
users' files
.br
.B \*9/ndb/auth
for mapping local userid to RADIUS userid
.SH SOURCE
.B \*9/src/cmd/auth/secstore
.SH SEE ALSO
.MR secstore 1


CHANGES
# Note: the name listed next to a change sometimes
# submitted the patch but more often just reported
# the problem.

# For entries more recent than January 30, 2008,
# see the Git repository history at 
# https://github.com/9fans/plan9port/.

January 30, 2008
	acme: fix regexp match choice bug
	devdraw: add F11 for full-screen toggle
	rio: accomodate full-screen apps more gracefully
	sam: fix regexp match choice bug

January 28, 2008
	xd: fix printing of 4-byte values on 64-bit machines (Sam Hopkins)

January 12, 2008
	delatex: reset lexer state at each beginning of file
	ipso: do not flush factotum keys unless using factotum file
	libdiskfs: fix ext2 directory reading bug (Mike Mammarella)
	lp: move devices to devices.sample
	rio: make full-screen work properly; add showevent
	venti: fewer calls to gettimeofday
	vmount: need to say -t nfs now on Linux

January 10, 2008
	acme: make STACK bigger for OS X (David Swasey)
	dist,unix: use MKSHELL=$PLAN9/bin/rc (John DeGood)
	libdraw: fix spinning event loop (David Bulkow)
	libregexp: fix match choice bug

December 7, 2007
	acme, sam, libregexp: regexp fixes and reversions

December 4, 2007
	hoc: print exact floating point output
	sam: turn off DEBUG logging (oops)
	vbackup: fix uninitialized variable use in -i code

November 28, 2007
	acme: refresh window properly after Tab command (Fazlul Shahriar)

November 27, 2007
	.hgignore: ignore more
	libdiskfs: make ext2 use little-endian disk format on big-endian machines (Joshua Wood)
	acme: accept clicks between body and scroll bar (Axel Belinfante)
	tpic: clean up better
	upas/smtp: remove intermediate files
	vbackup: add -i flag to avoid score prefetching (David Swasey)
	venti: add debugging HTML form

November 25, 2007
	acme: avoid infinite loop on search in 8125-byte file (Roger Peppe)
	lib9: avoid DIOCGMEDIASIZE on FreeBSD 4.11 (Rodrigo Readi)
	mailfs: use root in imapcopylist (Richard Bilson)
	moveplan9.files: add tmac/tmac.an (Tony Lainson)

November 23, 2007
	9pfuse: ignore Linux O_CLOEXEC flag (Michael Kaminsky)

November 22, 2007
	9pfuse: fix array bounds error (Michael Teichgräber)
	vbackup: fix fsysopen error message (Joshua Wood)

November 7, 2007
	9fs: use -n to mount sources (Jeff Sickel),
	9pfuse: MacFUSE moved again! (Jeff Sickel)

November 6, 2007
	.cvsignore: add 9pm (for old trees)
	unvf: ignore case in Content-Transfer-Encoding: line.
	validateattachment: reject malformed zip files

November 5, 2007
	building: use ar s flag to avoid ranlib (OS X)
	diff: continue after i/o errors when diffing many files
	factotum: silence gcc warning
	getns: turn / into _ in $DISPLAY
	libdiskfs: OS X linker fixes
	libmach: OS X linker fixes
	libmp: no text jump labels (OS X)
	libthread: use our own swapcontext on OS X
	mailfs: handle NIL body, add -r flag (Richard Bilson)

November 4, 2007
	man: document hg interface, fix CVS host name

November 3, 2007
	libthread: more OS X 10.5 changes (Jeff Sickel)

October 31, 2007
	venti: use 9ar, not ar (fixes Darwin build)

October 30, 2007
	libthread: OS X 10.5 tweaks (Jeff Sickel, Bakul Shah)

October 29, 2007
	grap: fclose, error context bug fixes (Doug Gwyn)
	venti: fix sync deadlock, add /proc stub

October 12, 2007
	9p: add -n to disable authentication
	9pserve: add -n to reject authentication
	srv: add -n to reject authentication

October 9, 2007
	mount: ask for Unix test explicitly (Chris King)

September 24-25, 2007
	venti: many fixes, new index cache

September 16, 2007
	factotum: fix build error (Axel Belinfante)

September 15, 2007
	auxstats: use acpi for battery on Linux (Andrey Mirtchovski)

September 13, 2007
	factotum: implement 9pcr (of dubious utility)

August 28, 2007
	9pserve: fix infinite loop bug

August 23, 2007
	venti: add -s flag, bug fixes (Michael Kaminsky)

August 22, 2007
	9.rc: change to work with rc -e (Tony Lainson)
	9pserve: fix races causing assert(c->nmsg == 0) to fail.
	acme: fix warning memory leak (Michael Teichgräber)
	libfmt: add fmtlocale to Makefile (Faried Nawaz)
	libthread: simplify alt tracking, possibly fixing bug
	mailfs: fix handling of FLAGS responses (Robert Vollmert)
	unix ports: Linux x86-64 Makefiles (Ævar Arnfjörð Bjarmason)

July 25, 2007
	9pfuse: even more MacFUSE path changes

June 29, 2007
	dict: look in $dictpath for dictionaries

June 24, 2007
	mailfs: fix subject line in header file.

June 23, 2007
	mailfs: various IMAP bug fixes
	Mail: add Get command, refresh at startup,
		fix deleted messages bug
	nedmail: refresh at startup, on i and y.

June 20, 2007
	9pfuse: add -a aname to get at sourcesdump

June 19, 2007
	9pfuse: the MacFUSE file locations changed again (Rob Pike)

June 18, 2007
	upas/marshal: remove debugging print
	9pserve: ignore window resizes (fix really should be elsewhere)
	libdiskfs: more HFS fixes (David Swasey)

June 15, 2007
	acme(1): fix Snarf bug, change Indent ON to apply to all windows.

June 12, 2007
	vcat: always write last block, to make disk correct size
	vbackup: more OS X fixes, add -M flag
	vbackup(8): document -M

June 9, 2007
	libregexp, sam, acme: day 1 bug in sam regexp code

June 8, 2007
	libdiskfs: add HFS resource fork support (David Swasey)

June 5, 2007
	ramfs: be less chatty

June 3, 2007
	vbackup: fix dotdot (David Swasey)

May 29, 2007
	libdiskfs: more HFS fixes (David Swasey)
	
May 28, 2007
	libdiskfs: missing Bterm, make ext2 quieter,
		add disk partitions, HFS code (David Swasey)

May 21, 2007
	libregexp: fix pointer comparison warning (Ramon de Vera)
	test: fix
	
May 20, 2007
	touch: print error if wstat fails
	libc: use futimes on Linux to implement dirfwstat

May 19, 2007
	test: import from Plan 9

May 18, 2007
	libregexp: fix bug about leading ^ in regexp on multiline text (Roger Peppe)

May 15, 2007
	libc: new media size code for _p9dir on FreeBSD (Lou Kamenov)
	acme(4): document tab width field in ctl file

May 10, 2007
	libc: more NaN fixes (rsc)
		use device names in _p9dir for FreeBSD (Lou Kamenov)

src/cmd/auth/secstore/secstored.c
#include <u.h>
#include <libc.h>
#include <bio.h>
#include <ndb.h>
#include <mp.h>
#include <libsec.h>
#include "SConn.h"
#include "secstore.h"

char *SECSTORE_DIR;
char* secureidcheck(char *, char *);   /* from /sys/src/cmd/auth/ */
extern char* dirls(char *path);

int verbose;
Ndb *db;

static void
usage(void)
{
	fprint(2, "usage: secstored [-R] [-S servername] [-s tcp!*!5356] [-v] [-x netmtpt]\n");
	exits("usage");
}

static int
getdir(SConn *conn, char *id)
{
	char *ls, *s;
	uchar *msg;
	int n, len;

	s = emalloc(Maxmsg);
	snprint(s, Maxmsg, "%s/store/%s", SECSTORE_DIR, id);

	if((ls = dirls(s)) == nil)
		len = 0;
	else
		len = strlen(ls);

	/* send file size */
	snprint(s, Maxmsg, "%d", len);
	conn->write(conn, (uchar*)s, strlen(s));

	/* send directory listing in Maxmsg chunks */
	n = Maxmsg;
	msg = (uchar*)ls;
	while(len > 0){
		if(len < Maxmsg)
			n = len;
		conn->write(conn, msg, n);
		msg += n;
		len -= n;
	}
	free(s);
	free(ls);
	return 0;
}

char *
validatefile(char *f)
{
	char *nl;

	if(f==nil || *f==0)
		return nil;
	if(nl = strchr(f, '\n'))
		*nl = 0;
	if(strchr(f,'/') != nil || strcmp(f,"..")==0 || strlen(f) >= 300){
		syslog(0, LOG, "no slashes allowed: %s\n", f);
		return nil;
	}
	return f;
}

static int
getfile(SConn *conn, char *id, char *gf)
{
	int n, gd, len;
	ulong mode;
	char *s;
	Dir *st;

	if(strcmp(gf,".")==0)
		return getdir(conn, id);

	/* send file size */
	s = emalloc(Maxmsg);
	snprint(s, Maxmsg, "%s/store/%s/%s", SECSTORE_DIR, id, gf);
	gd = open(s, OREAD);
	if(gd < 0){
		syslog(0, LOG, "can't open %s: %r\n", s);
		free(s);
		conn->write(conn, (uchar*)"-1", 2);
		return -1;
	}
	st = dirfstat(gd);
	if(st == nil){
		syslog(0, LOG, "can't stat %s: %r\n", s);
		free(s);
		conn->write(conn, (uchar*)"-1", 2);
		return -1;
	}
	mode = st->mode;
	len = st->length;
	free(st);
	if(mode & DMDIR) {
		syslog(0, LOG, "%s should be a plain file, not a directory\n", s);
		free(s);
		conn->write(conn, (uchar*)"-1", 2);
		return -1;
	}
	if(len < 0 || len > MAXFILESIZE){
		syslog(0, LOG, "implausible filesize %d for %s\n", len, gf);
		free(s);
		conn->write(conn, (uchar*)"-3", 2);
		return -1;
	}
	snprint(s, Maxmsg, "%d", len);
	conn->write(conn, (uchar*)s, strlen(s));

	/* send file in Maxmsg chunks */
	while(len > 0){
		n = read(gd, s, Maxmsg);
		if(n <= 0){
			syslog(0, LOG, "read error on %s: %r\n", gf);
			free(s);
			return -1;
		}
		conn->write(conn, (uchar*)s, n);
		len -= n;
	}
	close(gd);
	free(s);
	return 0;
}

static int
putfile(SConn *conn, char *id, char *pf)
{
	int n, nw, pd;
	long len;
	char s[Maxmsg+1];

	/* get file size */
	n = readstr(conn, s);
	if(n < 0){
		syslog(0, LOG, "remote: %s: %r\n", s);
		return -1;
	}
	len = atoi(s);
	if(len == -1){
		syslog(0, LOG, "remote file %s does not exist\n", pf);
		return -1;
	}else if(len < 0 || len > MAXFILESIZE){
		syslog(0, LOG, "implausible filesize %ld for %s\n", len, pf);
		return -1;
	}

	/* get file in Maxmsg chunks */
	if(strchr(pf,'/') != nil || strcmp(pf,"..")==0){
		syslog(0, LOG, "no slashes allowed: %s\n", pf);
		return -1;
	}
	snprint(s, Maxmsg, "%s/store/%s/%s", SECSTORE_DIR, id, pf);
	pd = create(s, OWRITE, 0660);
	if(pd < 0){
		syslog(0, LOG, "can't open %s: %r\n", s);
		return -1;
	}
	while(len > 0){
		n = conn->read(conn, (uchar*)s, Maxmsg);
		if(n <= 0){
			syslog(0, LOG, "empty file chunk\n");
			return -1;
		}
		nw = write(pd, s, n);
		if(nw != n){
			syslog(0, LOG, "write error on %s: %r", pf);
			return -1;
		}
		len -= n;
	}
	close(pd);
	return 0;

}

static int
removefile(SConn *conn, char *id, char *f)
{
	Dir *d;
	char buf[Maxmsg];

	snprint(buf, Maxmsg, "%s/store/%s/%s", SECSTORE_DIR, id, f);

	if((d = dirstat(buf)) == nil){
		snprint(buf, sizeof buf, "remove failed: %r");
		writerr(conn, buf);
		return -1;
	}else if(d->mode & DMDIR){
		snprint(buf, sizeof buf, "can't remove a directory");
		writerr(conn, buf);
		free(d);
		return -1;
	}

	free(d);
	if(remove(buf) < 0){
		snprint(buf, sizeof buf, "remove failed: %r");
		writerr(conn, buf);
		return -1;
	}
	return 0;
}

/* given line directory from accept, returns ipaddr!port */
static char*
remoteIP(char *ldir)
{
	int fd, n;
	char rp[100], ap[500];

	snprint(rp, sizeof rp, "%s/remote", ldir);
	fd = open(rp, OREAD);
	if(fd < 0)
		return strdup("?!?");
	n = read(fd, ap, sizeof ap);
	if(n <= 0 || n == sizeof ap){
		fprint(2, "error %d reading %s: %r\n", n, rp);
		return strdup("?!?");
	}
	close(fd);
	ap[n--] = 0;
	if(ap[n] == '\n')
		ap[n] = 0;
	return strdup(ap);
}

static int
dologin(int fd, char *S, int forceSTA)
{
	int i, n, rv;
	char *file, *mess;
	char msg[Maxmsg+1];
	PW *pw;
	SConn *conn;

	pw = nil;
	rv = -1;

	/* collect the first message */
	if((conn = newSConn(fd)) == nil)
		return -1;
	if(readstr(conn, msg) < 0){
		fprint(2, "remote: %s: %r\n", msg);
		writerr(conn, "can't read your first message");
		goto Out;
	}

	/* authenticate */
	if(PAKserver(conn, S, msg, &pw) < 0){
		if(pw != nil)
			syslog(0, LOG, "secstore denied for %s", pw->id);
		goto Out;
	}
	if((forceSTA || pw->status&STA) != 0){
		conn->write(conn, (uchar*)"STA", 3);
		if(readstr(conn, msg) < 10 || strncmp(msg, "STA", 3) != 0){
			syslog(0, LOG, "no STA from %s", pw->id);
			goto Out;
		}
		mess = secureidcheck(pw->id, msg+3);
		if(mess != nil){
			syslog(0, LOG, "secureidcheck denied %s because %s", pw->id, mess);
			goto Out;
		}
	}
	conn->write(conn, (uchar*)"OK", 2);
	syslog(0, LOG, "AUTH %s", pw->id);

	/* perform operations as asked */
	while((n = readstr(conn, msg)) > 0){
		syslog(0, LOG, "[%s] %s", pw->id, msg);

		if(strncmp(msg, "GET ", 4) == 0){
			file = validatefile(msg+4);
			if(file==nil || getfile(conn, pw->id, file) < 0)
				goto Err;

		}else if(strncmp(msg, "PUT ", 4) == 0){
			file = validatefile(msg+4);
			if(file==nil || putfile(conn, pw->id, file) < 0){
				syslog(0, LOG, "failed PUT %s/%s", pw->id, file);
				goto Err;
			}

		}else if(strncmp(msg, "RM ", 3) == 0){
			file = validatefile(msg+3);
			if(file==nil || removefile(conn, pw->id, file) < 0){
				syslog(0, LOG, "failed RM %s/%s", pw->id, file);
				goto Err;
			}

		}else if(strncmp(msg, "CHPASS", 6) == 0){
			if(readstr(conn, msg) < 0){
				syslog(0, LOG, "protocol botch CHPASS for %s", pw->id);
				writerr(conn, "protocol botch while setting PAK");
				goto Out;
			}
			pw->Hi = strtomp(msg, nil, 64, pw->Hi);
			for(i=0; i < 4 && putPW(pw) < 0; i++)
				syslog(0, LOG, "password change failed for %s (%d): %r", pw->id, i);
			if(i==4)
				goto Out;

		}else if(strncmp(msg, "BYE", 3) == 0){
			rv = 0;
			break;

		}else{
			writerr(conn, "unrecognized operation");
			break;
		}

	}
	if(n <= 0)
		syslog(0, LOG, "%s closed connection without saying goodbye\n", pw->id);

Out:
	freePW(pw);
	conn->free(conn);
	return rv;
Err:
	writerr(conn, "operation failed");
	goto Out;
}

void
main(int argc, char **argv)
{
	int afd, dfd, lcfd, forceSTA = 0;
	char adir[40], ldir[40], *remote;
	char *serve = "tcp!*!5356", *p, aserve[128];
	char *S = "secstore";
	char *dbpath;
	Ndb *db2;

	S = sysname();
	SECSTORE_DIR = unsharp("#9/secstore");
/*	setnetmtpt(net, sizeof(net), nil); */
	ARGBEGIN{
	case 'R':
		forceSTA = 1;
		break;
	case 's':
		serve = EARGF(usage());
		break;
	case 'S':
		S = EARGF(usage());
		break;
	case 'x':
		p = ARGF();
		if(p == nil)
			usage();
		USED(p);
	/*	setnetmtpt(net, sizeof(net), p); */
		break;
	case 'v':
		verbose++;
		break;
	default:
		usage();
	}ARGEND;

	if(!verbose)
		switch(rfork(RFNOTEG|RFPROC|RFFDG)) {
		case -1:
			sysfatal("fork: %r");
		case 0:
			break;
		default:
			exits(0);
		}

	snprint(aserve, sizeof aserve, "%s", serve);
	afd = announce(aserve, adir);
	if(afd < 0)
		sysfatal("%s: %r\n", aserve);
	syslog(0, LOG, "ANNOUNCE %s", aserve);
	for(;;){
		if((lcfd = listen(adir, ldir)) < 0)
			exits("can't listen");
		switch(rfork(RFPROC|RFFDG|RFNOWAIT)){
		case -1:
			fprint(2, "secstore forking: %r\n");
			close(lcfd);
			break;
		case 0:
			/* "/lib/ndb/common.radius does not exist" if db set before fork */
			db = ndbopen(dbpath=unsharp("#9/ndb/auth"));
			if(db == 0)
				syslog(0, LOG, "no ndb/auth");
			db2 = ndbopen(0);
			if(db2 == 0)
				syslog(0, LOG, "no ndb/local");
			db = ndbcat(db, db2);
			if((dfd = accept(lcfd, ldir)) < 0)
				exits("can't accept");
			alarm(30*60*1000); 	/* 30 min */
			remote = remoteIP(ldir);
			syslog(0, LOG, "secstore from %s", remote);
			free(remote);
			dologin(dfd, S, forceSTA);
			exits(nil);
		default:
			close(lcfd);
			break;
		}
	}
}
